tests/client_tests/node-postgres/README.rst

```diff
@@ -0,0 +1,107 @@
+====================================
+CrateDB client tests (node-postgres)
+====================================
+
+A nodejs_ application showcasing a test interaction with **CrateDB**.
+
+Staring the node-postgres_ module as middleware.
+
+The interaction:
+
+1. Opens a connection pool with CrateDB.
+2. Creates a table *logs*.
+3. Inserts 10000 log entries, which are recovered from a csv file
+4. Checks that all data have been loaded.
+5. Updates log entries whose object size is 0, to make it 40 for no
+   particular reason.
+6. Checks that the returning clause yields the updated data, with
+   the expected count of rows.
+7. Drops the table *logs*.
+8. Closes the connection pool.
+9. Exits the node process.
+
+Run it like:
+
+- `node app.js <host> <port>` (requires **CrateDB** to be running on <host>:<port>).
+- `./run.sh` (takes care of running **CrateDB** by means of cr8_).
+
+
+Sub modules
+-----------
+
+1. **app.js**: Entry point. Things to remark:
+
+   - Data are loaded from *resources/log_entries.csv*.
+   - Data are then inserted in batches of size 1000.
+     Beware, that means that we are producing strings:
+
+     `INSERT INTO t(..., ..., ...) VALUES (..., ..., ...), (..., ..., ...) ...`
+
+     that can be very large and eat up your process' mem, killing the node, depending
+     on the data morphology. Don't just take 1000 as gospel, make sure your node
+     can handle it for your data morphology.
+
+   - `expectedRowCount = wc -l *resources/log_entries.csv* minus one for the header`.
+   - `expectedUpdatedRowCount = grep ",0$" *resources/log_entries.csv* | wc -l`.
+   - At the end of the run, we report the time in millis it took to send the data
+     to **CrateDB**.
+
+2. **csv.js**: Exposes the `load` method that takes a callback expecting a
+   single parameter:
+
+   ::
+
+       (data) => {
+          // data.colNames
+          // data.length
+          // data.getValues(i where i in [0..data.length - 1])
+       };
+
+   *data* is a list of objects, each object is a row, each row has an attribute
+   per column, all rows contain the same attributes, although with different
+   values.
+
+   The method *getValues* is only available when *data* contains rows.
+   Returns an array containing the values for the particular row.
+
+   The attribute *colNames* is only available when *data* contains rows.
+   Returns an array containing the names of the columns.
+
+3. **db.js**: Contains and exposes methods used to interact with **CrateDB**.
+   Things to remark:
+
+   - We produce `data.length / batchSize` (10) multi-value insert requests.
+   - There are two types of execution, synchronous (to create the table),
+     and asynchronous, to carry out the upsert/drop.
+
+
+Docker
+------
+
+Our build system runs the **node-postgres** client tests on a docker_
+container, with Dockerfile:
+
+::
+
+    FROM openjdk:13
+
+    RUN \
+      yum install deltarpm -y && \
```

Comment:
Please avoid duplicating information that will quickly get outdated. Link to the Dockerfile if you think it is necessary, but people can just as well read the Jenkinsfile and they'll have the assurance that they read the up to date truth.

Reply:
indeed!, I had not thought about that, will put a link.